libnavigation-core/src/main/java/com/mapbox/navigation/core/reroute/RerouteController.kt

```diff
@@ -0,0 +1,78 @@
+package com.mapbox.navigation.core.reroute
+
+import com.mapbox.api.directions.v5.models.DirectionsRoute
+import com.mapbox.navigation.core.MapboxNavigation
+import com.mapbox.navigation.core.trip.session.OffRouteObserver
+
+/**
+ * Reroute controller allow change reroute logic externally. Use [MapboxNavigation.setRerouteController]
+ * to replace that logic.
+ */
+interface RerouteController {
```

Comment:
Let's add a `RerouteStateObserver` that a developer could subscribe to update the UI (for example).

Reply:
developer already has callback to keep user updated that he's out of route, it's `OffRouteObserver`. basically this state extend offroute one. 
Moreover, I'd like to remove `val rerouteState: RerouteState` from interface and use it internal only.

Reply:
That's a different use case though. The `RerouteState` introduced here gives us an opportunity to update UI based on the request state, and also provide details or retry when it fails. The `OffRouteObserver` doesn't have enough information to do that.

Reply:
Yeah, form this perspective of view right, and I'm totally agree. From the other hand we're gonna to production very soon, and expose smth easier than hide in that case. I'd prefer expose it late when we investigate problems through using that API internally

Reply:
That's fair. The tradeoff here is that adding a `RerouteStateObserver` to the interface later would be a breaking change.

Reply:
should choose lesser of two evils 😈  
What do you think if we expose it, add `RerouteStateObserver` and 1tap will start to use it? Before end of July we'll make a good decision

Reply:
Sounds good to me :+1:

Reply:
We can always expose it "publicly" under the `internal` package and experiment - making clear to other clients that it's meant to be used internally only.

Reply:
but we still need to have `public` `setters`/`getters` in `MapboxNavigation` in particular case, so it's not prevent  to use internal-package classes

Reply:
Same question as above
> Wondering if with the addition of `RerouteController.RerouteStateObserver` we can remove the `OffRouteObserver` / if `OffRouteObserver` is useless / can be replaced by `RerouteStateObserver` altogether 🤔 What do you think?

cc @kmadsen @LukasPaczos

Reply:
That's a good point, however, I'm not sure whether having a **reroute** controller the only source of truth for the **off route** event is going to be clear. We might be off route and never want to reroute.

Reply:
Also, the reroute state is more about a download progress state than a positioning state.

Reply:
@LukasPaczos totally agree, my 5 cents https://github.com/mapbox/mapbox-navigation-android/pull/3184#discussion_r450118294

Reply:
Asked because we've heard complaints around https://github.com/mapbox/mapbox-navigation-android/blob/907899a90eb70921641a9b6d6efe2cdcfc79d77f/libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/OffRouteObserver.kt#L5-L14 as only pushing back a flag `offRoute` is not enough and would be great to have more info e.g. in telemetry we need to listen for re-routes via `OffRouteObserver` AND if `true` catch right after the first callback from `RoutesObserver` to update the events accordingly which is kinda complicated and we were wondering if we could push the route chosen back somehow making the API easier to use and consume and provide to clients whole state for when an off-route happens. In any case, we can revisit this as part of a different PR. Could you cut a quick ticket to keep track of this discussion?

Reply:
catched in ticket https://github.com/mapbox/mapbox-navigation-android/issues/3292